Feat(web-react): Introduce `Flex` component (#1594)
lmc-eu/spirit-design-system: packages/web-react/src/components/Flex/README.md

# Flex

Flex is a component that allows you to create a flexible one-dimensional layout.

## Basic Usage

Row layout:

```jsx
<Flex>
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</Flex>
```

Column layout:

```jsx
<Flex direction="column">
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</div>
```

Usage with a list:

```jsx
<Flex elementType="ul" direction="column">
  <li>Item 1</li>
  <li>Item 2</li>
  <li>Item 3</li>
</Flex>
```

ℹ️ For the row layout, the Flex component uses the [`display: flex`][mdn-display-flex] CSS property. For the column
layout, [`display: grid`][mdn-display-grid] is used because of technical advantages: better overflow control or
alignment API consistency.

## Responsive Direction

> To create a responsive layout, pass a object as the value for the direction property, using breakpoint keys to specify different layouts for each screen size.

Reviewer: ```suggestion
To create a responsive layout, pass an object as the value for the `direction` property, using breakpoint keys to specify different layouts for each screen size.
```

```jsx
<Flex direction={{ mobile: 'column', tablet: 'row' }}>
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</Flex>
```

## Wrapping

By default, Flex items will not wrap. To enable wrapping on all breakpoints, use the
`isWrapping` prop.

```jsx
<Flex isWrapping>
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</Flex>
```

### Responsive Wrapping

> To create a responsive layout, pass a object as the value for the `isWrapping` property, using breakpoint keys to specify different layouts for each screen size.

Reviewer: ```suggestion
To create a responsive wrapping layout, pass an object as the value for the `isWrapping` property, using breakpoint keys to specify different wrapping for each screen size.
```

```jsx
<Flex isWrapping={{ mobile: true, tablet: false }}>
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</Flex>
```

## Alignment

### Horizontal Alignment

Flex can be horizontally aligned as stretched (default), or justified to the left, center, or right. Additionally, you
can evenly distribute the items using the space-between value. These values come from the extended

> [alignmentX dictionary][dictionary-alignment]. Using a corresponding prop will align the Flex items accordingly:

Reviewer: > accordingly:

what?

### Vertical Alignment

Similarly to the horizontal alignment, Flex can be vertically aligned as stretched (default), or justified to the top,
center, or bottom. There is also an option to align the items to the baseline. These values come from the extended
[alignmentY dictionary][dictionary-alignment]. Using a corresponding prop will align the Flex items accordingly:

Example:

```jsx
<Flex alignmentX="right" alignmentY="baseline">
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</div>
```

### Responsive Alignment

> To create a responsive layout, pass a object as the value for the property, using breakpoint keys to specify different layouts for each screen size.

Reviewer: ```suggestion
To create a responsive alignment, pass an object as the value for the property, using breakpoint keys to specify different alignments for each screen size.
```

  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</Flex>
```

## Custom Spacing

You can use the `spacing` prop to apply custom spacing between items in both horizontal and vertical directions. The prop
accepts either a spacing token (e.g. `space-100`) or an object with breakpoint keys and spacing token values.

Custom spacing:

```jsx
<Flex spacing="space-1200">
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</div>
```

Custom responsive spacing:

```jsx
<Flex spacing={{ mobile: 'space-400', tablet: 'space-800' }}>
  <div>Item 1</div>
  <div>Item 2</div>
  <div>Item 3</div>
</Flex>
```

## API

| Name          | Type                                                                | Default   | Required | Description                                                                                                                             |
| ------------- | ------------------------------------------------------------------- | --------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------- |

> | `alignmentX`  | [[AlignmentXExtended dictionary][alignment-dictionary] \| `object`] | `stretch` | ✕        | Apply horizontal alignment of items, use object to set responsive values, e.g. `{ mobile: 'left', tablet: 'center', desktop: 'right' }` |

Reviewer: ```suggestion
| `alignmentX`  | [[AlignmentXExtended dictionary][alignment-dictionary] \| `object`] | `stretch` | ✕        | Apply horizontal alignment of items, use an object to set responsive values, e.g. `{ mobile: 'left', tablet: 'center', desktop: 'right' }` |
```

> | `alignmentY`  | [[AlignmentYExtended dictionary][alignment-dictionary] \| `object`] | `stretch` | ✕        | Apply vertical alignment of items, use object to set responsive values, e.g. `{ mobile: 'top', tablet: 'center', desktop: 'bottom' }`   |

Reviewer: ```suggestion
| `alignmentY`  | [[AlignmentYExtended dictionary][alignment-dictionary] \| `object`] | `stretch` | ✕        | Apply vertical alignment of items, use an object to set responsive values, e.g. `{ mobile: 'top', tablet: 'center', desktop: 'bottom' }`   |
```

> | `direction`   | [`row` \| `column`]                                                 | `row`     | ✕        | Direction of the items, use object to set responsive values, e.g. `{ mobile: 'row', tablet: 'row', desktop: 'column' }`                 |

Reviewer: add object to type?

> | `direction`   | [`row` \| `column`]                                                 | `row`     | ✕        | Direction of the items, use object to set responsive values, e.g. `{ mobile: 'row', tablet: 'row', desktop: 'column' }`                 |

Reviewer: ```suggestion
| `direction`   | [`row` \| `column`]                                                 | `row`     | ✕        | Direction of the items, use an object to set responsive values, e.g. `{ mobile: 'row', tablet: 'row', desktop: 'column' }`                 |
```

| `elementType` | HTML element                                                        | `div`     | ✕        | Element type to use for the Grid                                                                                                        |

> | `isWrapping`  | `bool`                                                              | `false`   | ✕        | Whether items will wrap, use object to set responsive values, e.g. `{ mobile: true, tablet: true, desktop: false }`                     |

Reviewer: add object to type?

> | `isWrapping`  | `bool`                                                              | `false`   | ✕        | Whether items will wrap, use object to set responsive values, e.g. `{ mobile: true, tablet: true, desktop: false }`                     |

Reviewer: ```suggestion
| `isWrapping`  | `bool`                                                              | `false`   | ✕        | Whether items will wrap, use an object to set responsive values, e.g. `{ mobile: true, tablet: true, desktop: false }`                     |
```

> | `spacing`     | [`SpaceToken` \| `Partial<Record<BreakpointToken, SpaceToken>>`]    | —         | ✕        | Apply [custom spacing](#custom-spacing) in both horizontal and vertical directions between items                                           |

Reviewer 2: ```suggestion
| `spacing`     | \[`SpaceToken` \| `Partial<Record<BreakpointToken, SpaceToken>>`]    | —         | ✕        | Apply [custom spacing](#custom-spacing) in both horizontal and vertical directions between items                                           |
```

Please use a backslash at the beginning of these strings. The Markdown can misinterpret this as a reference. For more, please see https://github.com/lmc-eu/spirit-design-system/pull/1587

Author: added
